### PR TITLE
Add spawn scheduler docs

### PR DESCRIPTION
**Pull request by mohanson**
Please check whether the description of scheduler is correct.

### CODE REVIEW

File: src/syscalls/native.rs [hunk 1]
```diff
@@ -590,14 +590,53 @@ pub struct SpawnArgs {
...
+/// This document describes the design and functionality of a scheduler algorithm, covering process states, system call
+/// behavior, message handling, and priority rules. The scheduler manages virtual processes, transitioning them through
+/// various states based on operations and interactions, to ensure efficient scheduling.
+///
+/// # Thread States
+///
+/// Each process within this scheduler has one of the following five states:
+/// * Runnable: The process is ready to execute.
+/// * Terminated: The process has completed its execution.
+/// * Wait: The process is waiting for another process to exit before it can continue.
```
Comment (xxuejie) [inline src/syscalls/native.rs]: I suggest renaming this to `WaitForExit` to be more precise

File: src/syscalls/native.rs [hunk 2]
```diff
@@ -590,14 +590,53 @@ pub struct SpawnArgs {
...
+///
+/// # Scheduler Algorithm
+///
+/// This document describes the design and functionality of a scheduler algorithm, covering process states, system call
+/// behavior, message handling, and priority rules. The scheduler manages virtual processes, transitioning them through
+/// various states based on operations and interactions, to ensure efficient scheduling.
+///
+/// # Thread States
+///
+/// Each process within this scheduler has one of the following five states:
```
Comment (xxuejie) [inline src/syscalls/native.rs]: Just throwing something out: is `Running` also a state?

File: src/syscalls/native.rs [hunk 3]
```diff
@@ -590,14 +590,53 @@ pub struct SpawnArgs {
...
+/// Specific system calls are responsible for changing the state of a process within this scheduler:
+/// * spawn: Creates a new process, initializing it in the Runnable state.
+/// * read: Attempts to read data from a file descriptor. If data is unavailable, the process state changes to
+///     WaitForRead.
+/// * write: Attempts to write data to a file descriptor. If the operation is blocked due to data needing to be read by
+///     another process, the process enters the WaitForWrite state.
+/// * wait: Waits for a target process to exit. Once the target process has terminated, the waiting process transitions
+///     to Runnable.
+///
+/// # Message Handling and State Recovery
```
Comment (xxuejie) [inline src/syscalls/native.rs]: I'm not sure if we should introduce `message handling` here, what is described here is simply IOs, there are no messages yet.

File: src/syscalls/native.rs [hunk 4]
```diff
@@ -590,14 +590,53 @@ pub struct SpawnArgs {
...
+/// Note: available after ckb 2nd hardfork.
 ///
 /// # Arguments
 ///
 /// * `index`, `source`, `bounds` and `place` - same as exec.
 /// * `spgs` - spawn arguments.
 ///
 /// Returns success or a syscall error.
+///
+/// # Scheduler Algorithm
```
Comment (xxuejie) [inline src/syscalls/native.rs]: Better name this to scheduler algorithm v1, as latter hardforks might introduce v2, v3, et.

File: src/syscalls/native.rs (hunk 2 above)
Comment (mohanson) [reply src/syscalls/native.rs]: From the code, it seems that `Running` should not be a state

https://github.com/nervosnetwork/ckb/blob/develop/script/src/types.rs#L394-L410

Comment (xxuejie) [reply src/syscalls/native.rs]: Yes but the code in CKB is the actual implementation, the `VmState` structure in implementation, does not always mean a conceptual model. I'm raising a different question: from a higher level, when one is using `ckb-std` to construct apps on CKB, is running, treated as a state of a program?

Looking at literature of OSes: running is treated as a state: https://www.geeksforgeeks.org/states-of-a-process-in-operating-systems/

Comment (mohanson) [reply src/syscalls/native.rs]: Added the Running state, and added information on how to enter and exit the Running state at the end.

File: src/syscalls/native.rs [hunk 5]
```diff
@@ -590,14 +590,55 @@ pub struct SpawnArgs {
...
+/// * Terminated: The process has completed its execution.
+/// * WaitForRead: The process is waiting for data to be available for it to read.
+/// * WaitForWrite: The process is waiting for another process to read data it wants to write.
+/// * WaitForExit: The process is waiting for another process to exit before it can continue.
+///
+/// # System Calls and State Transitions
+///
+/// Specific system calls are responsible for changing the state of a process within this scheduler:
+/// * spawn: Creates a new process, initializing it in the Runnable state.
+/// * read: Attempts to read data from a file descriptor. If data is unavailable, the process state changes to
```
Comment (XuJiandong) [inline src/syscalls/native.rs]: Add more description when data is less than expected.

File: src/syscalls/native.rs [hunk 6]
```diff
@@ -590,14 +590,55 @@ pub struct SpawnArgs {
...
+///     WaitForRead.
+/// * write: Attempts to write data to a file descriptor. If the operation is blocked due to data needing to be read by
```
Comment (XuJiandong) [inline src/syscalls/native.rs:624]: Add description that all data should be consumed

File: src/syscalls/native.rs [hunk 7]
```diff
@@ -590,14 +590,55 @@ pub struct SpawnArgs {
...
+/// * A WaitForRead process becomes Runnable once the needed data has been read successfully.
+/// * A WaitForWrite process transitions to Runnable once its data has been successfully read by another process.
+///
+/// # Process Priority
+///
+/// The scheduler assigns incremental IDs to processes, establishing an execution order:
+/// * The root process has an ID of 0.
+/// * When multiple processes are in the Runnable state, the scheduler selects the process with the lowest ID to execute
+///     first. This ensures a predictable and fair ordering of execution for processes ready to run. The selected
+///     process status is changed to Running, and it continues to run until its status is changed.
```
Comment (XuJiandong) [inline src/syscalls/native.rs:642]: Add links to deadlock in RFCs.

**Review by XuJiandong — CHANGES_REQUESTED** [review]

File: src/syscalls/native.rs (hunk 6 above)
Comment (mohanson) [reply src/syscalls/native.rs:624]: Descriptions are given in the IO handling section.

File: src/syscalls/native.rs (hunk 7 above)
Comment (mohanson) [reply src/syscalls/native.rs:642]: The PR about deadlock has not yet been merged.

File: src/syscalls/native.rs (hunk 6 above)
Comment (XuJiandong) [reply src/syscalls/native.rs:624]: This description is insufficient. On Linux and other platforms, the `write` operation may perform a partial write, only writing some of the data. In our design, all data is written completely.

**Review by XuJiandong — CHANGES_REQUESTED** [review]

**Review by xxuejie — APPROVED** [review]

**Pull request merged**